### PR TITLE
pkg/userfunctions: Add user-defined functions to ConTest

### PR DESCRIPTION
**Pull request by walterchris**
Slightly Restructure plugins.go to work with externally defined
user-defined functions. This approach is close to what we use with
plugins. To make it 100% compatible with plugins, I would also need to
touch pluginRegistry itself.

Signed-off-by: Christian Walter <christian.walter@9elements.com>

### CODE REVIEW

**Comment by tfg13** [comment]
Curious, why is it different from the way the plugins work? Just wondering :)

**Comment by walterchris** [comment]
Plugins do all get registered within the pluginRegistry. In plugins.go we do _load_ them, that means that we do build up a map of Name and Plugin Instances on which we do later call the functions on (RunReport, etc.).

User-defined functions only needs to be registered within test package. We do build up a string -> function map which will be applied as Go Templates to the JSON Input e.g. Test definitions.

I think we could also probably add udf to the pluginRegistry and let the registerFunction function call this. I do not see major improvements/benefits on either ways.

**Comment by tfg13** [comment]
Another thing re:`UserDefinedFunction` - is the `Name()` function used anywhere? If not feel free to throw it out :)

File: pkg/userfunctions/ocp/ocp.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+var userFunctions = map[string]interface{}{
+	"getPrivateKeyFile": func(a ...string) (string, error) {
+		if len(a) == 0 {
+			return "", errors.New("hostparam: no arg specified")
+		}
+		if len(a) > 1 {
+			return "", errors.New("hostparam: too many parameters")
+		}
```
[the remark this answers is not in the document]
Comment (walterchris) [reply pkg/userfunctions/ocp/ocp.go:38]: Ack.

File: pkg/userfunctions/userdefinedfunction.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+package userdefinedfunction
+
+// UserDefinedFunction Interface - Every UDF needs to implement at least these
+// functions in order to get hook up within ConTest.
+type UserDefinedFunction interface {
```
[the remark this answers is not in the document]
Comment (walterchris) [reply pkg/userfunctions/userdefinedfunction.go:10]: Ah no.

I wanted to work out something like plugins.. looks like I did not. I just remove this.

**Comment by walterchris** [comment]
Delete Name completely..

I adopted this from the plugins - but that was never used anywhere. :)

**Review by tfg13 — APPROVED** [review]

**Pull request merged**